Version assertion

MoonShiesty opened the pull request:
# Goals of PR

Version assertion data, which currently is currently equivalent to vmHash.

- Describe changes here

Prepend a version byte to assertion data, and store as a blob and modify validator to publish new versioned assertion data blob format.

- Write notes here

These changes increase the size of `Rollup.sol` over the transaction size limit (by ~100 with optimizer on and 1 run). We'll have to split up `Rollup.sol` to deploy these changes.

The decoding versioned assertion data cannot revert or validators will be unable to challange assertions with invalid encoding.

MoonShiesty commented on contracts/src/Rollup.sol:
```diff
@@ -57,9 +57,12 @@ abstract contract RollupBase is
     struct AssertionState {
         mapping(address => bool) stakers; // all stakers that have ever staked on this assertion.
-        mapping(bytes32 => bool) childStateHashes; // child assertion vm hashes
+        mapping(bytes => bool) childAssertionData; // child assertion data
```
not sure if the key to this mapping should be the versioned blob, or leave it as a state hash

chalkandpaste replied:
I think including the version is probably better, but the map may cost us more because of being arbitrary length byte array? @LEAFERx

MoonShiesty replied:
this is reverted now, based on changes suggested by @ujvl

chalkandpaste commented on contracts/src/IRollup.sol:
```diff
@@ -308,10 +314,10 @@ interface IRollup {
-     * @param vmHash New VM hash.
-     * @param inboxSize Size of inbox corresponding to assertion (number of transactions).
+     * @param blockNum Block number this assertion advances to.
+     * @param assertionData Versioned assertion data
```
Can you leave a little bit of detail on what the structure of the data is. Maybe a ref somewhere or a "version int8 | vmHash int256" (assuming that's the structure). Maybe it's not terribly important as the data doesn't get used.

MoonShiesty commented:
@ujvl ready for review. I changed the implementation to version the `vmHash`, by encoding the assertion `vmHash`, as `StateRoot = keccak(version | vmHash)`, instead of using a blob data format.

I renamed the `vmHash` data to `StateRoot` because in the future version it might contain additional hashed data. Optimism/Magi calls it `OutputRoot`, not sure what the best name is.

ujvl commented on services/sidecar/rollup/services/validator/stateRoot.go:
filename should be lowercase with underscore

ujvl commented on services/sidecar/rollup/types/assertion.go:
```diff
@@ -2,14 +2,14 @@ package types
 type Assertion struct {
 	ID         *big.Int
-	VmHash     common.Hash
+	StateRoot  Bytes32
```
name conflict with ethereum state root, we should rename this

MoonShiesty commented:
@ujvl ready for review - renamed to `state_commitment` and squashed into a single commit

ujvl commented on services/sidecar/rollup/services/validator/validator.go:
```diff
@@ -34,8 +35,8 @@ type Validator struct {
-	l2BlockNum uint64
-	l2VMHash   common.Hash
+	l2BlockNum      uint64
+	stateCommitment specularTypes.Bytes32
```
nit for clarity
```suggestion
	l2StateCommitment specularTypes.Bytes32
```

MoonShiesty replied:
done

ujvl commented on services/sidecar/rollup/services/validator/state_commitment.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+// StateCommitment returns the keccak256 hash of the marshaled L2 state commitment
+func StateCommitment(stateCommitment VersionedStateCommitment) types.Bytes32 {
```
nit: consider `ToBytes32()` maybe
